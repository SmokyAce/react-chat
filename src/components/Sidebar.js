import React from 'react';
import { withStyles } from 'material-ui/styles';
import PropTypes from 'prop-types';
import Drawer from 'material-ui/Drawer';
import Divider from 'material-ui/Divider';
import TextField from 'material-ui/TextField';
import BottomNavigation, { BottomNavigationAction } from 'material-ui/BottomNavigation';
import RestoreIcon from 'material-ui-icons/Restore';
import ExploreIcon from 'material-ui-icons/Explore';
import IconButton from 'material-ui/IconButton';
import ChevronLeftIcon from 'material-ui-icons/ChevronLeft';

import ChatList from './ChatList';
import NewChatButton from './NewChatButton';

const styles = theme => ({
  drawerPaper: {
    position: 'relative',
    height: '100%',
    width: theme.drawerWidth,
  },
  drawerHeader: {
    ...theme.mixins.toolbar,
    display: 'flex',
    paddingLeft: theme.spacing.unit * 3,
    // paddingRight: theme.spacing.unit * 3,
  },
  drawerCloseButton: {
    alignSelf: 'center',
  },
});

class Sidebar extends React.Component {
  static propTypes = {
    classes: PropTypes.objectOf(PropTypes.string).isRequired,
    chats: PropTypes.shape({
      my: PropTypes.array.isRequired,
      all: PropTypes.array.isRequired,
      active: PropTypes.object,
    }).isRequired,
    actions: PropTypes.shape({
      createChat: PropTypes.func.isRequired,
    }).isRequired,
    isConnected: PropTypes.bool.isRequired,
    open: PropTypes.bool.isRequired,
  };

  state = {
    activeAction: 0,
    searchValue: '',
  };

  handleChangeActiveAction = (e, value) => this.setState({ activeAction: value });

  handleSearchValueChange = event =>
    this.setState({
      searchValue: event.target.value,
    });

  filterChats = chats =>
    chats
      .filter(chat => chat.title.toLowerCase().includes(this.state.searchValue.toLowerCase()))
      .sort((one, two) => (one.title.toLowerCase() <= two.title.toLowerCase() ? -1 : 1));

  render() {
    const {
      classes, chats, actions, isConnected, open,
    } = this.props;
    const { activeAction, searchValue } = this.state;

    return (
      <Drawer
        variant='persistent'
        open={open}
        classes={{
          paper: classes.drawerPaper,
        }}
      >
        <div className={classes.drawerHeader}>
          <TextField
            fullWidth
            margin='normal'
            placeholder='Search chats...'
            onChange={this.handleSearchValueChange}
            value={searchValue}
          />
          <IconButton className={classes.drawerCloseButton} onClick={actions.closeSidebar}>
            <ChevronLeftIcon />
          </IconButton>
        </div>
        <Divider />
        <ChatList
          chats={this.filterChats(activeAction === 0 ? chats.my : chats.all)}
          activeChat={chats.active}
          setActiveChat={actions.setActiveChat}
          disabled={!isConnected}
        />
        <NewChatButton createChat={actions.createChat} disabled={!isConnected} />
        <BottomNavigation
          showLabels
          value={this.state.activeAction}
          onChange={this.handleChangeActiveAction}
        >
          <BottomNavigationAction label='My Chats' icon={<RestoreIcon />} />
          <BottomNavigationAction label='Explore' icon={<ExploreIcon />} />
        </BottomNavigation>
      </Drawer>
    );
  }
}

export default withStyles(styles)(Sidebar);
