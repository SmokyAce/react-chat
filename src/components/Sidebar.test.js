/* eslint-env jest */
import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import Sidebar from './Sidebar';

jest.mock('./ChatList', () => () => 'ChatList');
jest.mock('./NewChatButton', () => () => 'NewChatButton');

global.matchMedia = () => ({
  addListener: jest.fn(),
  removeListener: jest.fn(),
});

const mockProps = {
  chats: {
    my: [
      {
        _id: '1',
        title: 'test',
        createAt: '2018-03-08T13:57:58.706Z',
      },
    ],
    all: [
      {
        _id: '1',
        title: 'test',
        createAt: '2018-03-08T13:57:58.706Z',
      },
    ],
    active: {
      _id: '1',
      title: 'test',
    },
  },
  actions: {
    createChat: jest.fn(),
    openSidebar: jest.fn(),
    closeSidebar: jest.fn(),
  },
  isConnected: false,
  open: true,
};

describe('<Sidebar />', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<Sidebar {...mockProps} />, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it('renders correctly', () => {
    const tree = renderer.create(<Sidebar {...mockProps} />).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('renders correctly isConnected', () => {
    const tree = renderer.create(<Sidebar {...mockProps} isConnected />).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
