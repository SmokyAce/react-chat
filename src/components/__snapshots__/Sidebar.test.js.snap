// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Sidebar /> renders correctly 1`] = `
<div
  className="MuiDrawer-root-116 MuiDrawer-docked-117"
  style={
    Object {
      "width": 0,
    }
  }
>
  <div
    className="MuiPaper-root-128 MuiPaper-elevation0-130 MuiDrawer-paper-118 Sidebar-drawerPaper-113 MuiDrawer-paperAnchorLeft-119 MuiDrawer-paperAnchorDockedLeft-123"
    style={
      Object {
        "pointerEvents": "",
      }
    }
  >
    <div
      className="Sidebar-drawerHeader-114"
    >
      <div
        className="MuiFormControl-root-155 MuiFormControl-marginNormal-156 MuiFormControl-fullWidth-158"
      >
        <div
          className="MuiInputBase-root-172 MuiInput-root-159 MuiInput-underline-163 MuiInputBase-fullWidth-181 MuiInput-fullWidth-166 MuiInputBase-formControl-173 MuiInput-formControl-160"
          onClick={[Function]}
        >
          <input
            aria-invalid={false}
            className="MuiInputBase-input-182 MuiInput-input-167"
            disabled={false}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            placeholder="Search chats..."
            required={false}
            type="text"
            value=""
          />
        </div>
      </div>
      <button
        className="MuiButtonBase-root-195 MuiIconButton-root-189 Sidebar-drawerCloseButton-115"
        disabled={false}
        onBlur={[Function]}
        onClick={[MockFunction]}
        onContextMenu={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex="0"
        type="button"
      >
        <span
          className="MuiIconButton-label-194"
        >
          <svg
            aria-hidden="true"
            className="MuiSvgIcon-root-198"
            focusable="false"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <path
              d="M15.41 7.41L14 6l-6 6 6 6 1.41-1.41L10.83 12z"
            />
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
          </svg>
        </span>
        <span
          className="MuiTouchRipple-root-218"
        />
      </button>
    </div>
    <hr
      className="MuiDivider-root-207"
    />
    ChatList
    NewChatButton
    <div
      className="MuiBottomNavigation-root-212"
    >
      <button
        className="MuiButtonBase-root-195 MuiBottomNavigationAction-root-213 MuiBottomNavigationAction-selected-214"
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex="0"
        type="button"
      >
        <span
          className="MuiBottomNavigationAction-wrapper-216"
        >
          <svg
            aria-hidden="true"
            className="MuiSvgIcon-root-198"
            focusable="false"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
            <path
              d="M13 3c-4.97 0-9 4.03-9 9H1l3.89 3.89.07.14L9 12H6c0-3.87 3.13-7 7-7s7 3.13 7 7-3.13 7-7 7c-1.93 0-3.68-.79-4.94-2.06l-1.42 1.42C8.27 19.99 10.51 21 13 21c4.97 0 9-4.03 9-9s-4.03-9-9-9zm-1 5v5l4.28 2.54.72-1.21-3.5-2.08V8H12z"
            />
          </svg>
          <span
            className="MuiBottomNavigationAction-label-217 MuiBottomNavigationAction-selected-214"
          >
            My Chats
          </span>
        </span>
        <span
          className="MuiTouchRipple-root-218"
        />
      </button>
      <button
        className="MuiButtonBase-root-195 MuiBottomNavigationAction-root-213"
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex="0"
        type="button"
      >
        <span
          className="MuiBottomNavigationAction-wrapper-216"
        >
          <svg
            aria-hidden="true"
            className="MuiSvgIcon-root-198"
            focusable="false"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 10.9c-.61 0-1.1.49-1.1 1.1s.49 1.1 1.1 1.1c.61 0 1.1-.49 1.1-1.1s-.49-1.1-1.1-1.1zM12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm2.19 12.19L6 18l3.81-8.19L18 6l-3.81 8.19z"
            />
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
          </svg>
          <span
            className="MuiBottomNavigationAction-label-217"
          >
            Explore
          </span>
        </span>
        <span
          className="MuiTouchRipple-root-218"
        />
      </button>
    </div>
  </div>
</div>
`;

exports[`<Sidebar /> renders correctly isConnected 1`] = `
<div
  className="MuiDrawer-root-116 MuiDrawer-docked-117"
  style={
    Object {
      "width": 0,
    }
  }
>
  <div
    className="MuiPaper-root-128 MuiPaper-elevation0-130 MuiDrawer-paper-118 Sidebar-drawerPaper-113 MuiDrawer-paperAnchorLeft-119 MuiDrawer-paperAnchorDockedLeft-123"
    style={
      Object {
        "pointerEvents": "",
      }
    }
  >
    <div
      className="Sidebar-drawerHeader-114"
    >
      <div
        className="MuiFormControl-root-155 MuiFormControl-marginNormal-156 MuiFormControl-fullWidth-158"
      >
        <div
          className="MuiInputBase-root-172 MuiInput-root-159 MuiInput-underline-163 MuiInputBase-fullWidth-181 MuiInput-fullWidth-166 MuiInputBase-formControl-173 MuiInput-formControl-160"
          onClick={[Function]}
        >
          <input
            aria-invalid={false}
            className="MuiInputBase-input-182 MuiInput-input-167"
            disabled={false}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            placeholder="Search chats..."
            required={false}
            type="text"
            value=""
          />
        </div>
      </div>
      <button
        className="MuiButtonBase-root-195 MuiIconButton-root-189 Sidebar-drawerCloseButton-115"
        disabled={false}
        onBlur={[Function]}
        onClick={[MockFunction]}
        onContextMenu={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex="0"
        type="button"
      >
        <span
          className="MuiIconButton-label-194"
        >
          <svg
            aria-hidden="true"
            className="MuiSvgIcon-root-198"
            focusable="false"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <path
              d="M15.41 7.41L14 6l-6 6 6 6 1.41-1.41L10.83 12z"
            />
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
          </svg>
        </span>
        <span
          className="MuiTouchRipple-root-218"
        />
      </button>
    </div>
    <hr
      className="MuiDivider-root-207"
    />
    ChatList
    NewChatButton
    <div
      className="MuiBottomNavigation-root-212"
    >
      <button
        className="MuiButtonBase-root-195 MuiBottomNavigationAction-root-213 MuiBottomNavigationAction-selected-214"
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex="0"
        type="button"
      >
        <span
          className="MuiBottomNavigationAction-wrapper-216"
        >
          <svg
            aria-hidden="true"
            className="MuiSvgIcon-root-198"
            focusable="false"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
            <path
              d="M13 3c-4.97 0-9 4.03-9 9H1l3.89 3.89.07.14L9 12H6c0-3.87 3.13-7 7-7s7 3.13 7 7-3.13 7-7 7c-1.93 0-3.68-.79-4.94-2.06l-1.42 1.42C8.27 19.99 10.51 21 13 21c4.97 0 9-4.03 9-9s-4.03-9-9-9zm-1 5v5l4.28 2.54.72-1.21-3.5-2.08V8H12z"
            />
          </svg>
          <span
            className="MuiBottomNavigationAction-label-217 MuiBottomNavigationAction-selected-214"
          >
            My Chats
          </span>
        </span>
        <span
          className="MuiTouchRipple-root-218"
        />
      </button>
      <button
        className="MuiButtonBase-root-195 MuiBottomNavigationAction-root-213"
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex="0"
        type="button"
      >
        <span
          className="MuiBottomNavigationAction-wrapper-216"
        >
          <svg
            aria-hidden="true"
            className="MuiSvgIcon-root-198"
            focusable="false"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 10.9c-.61 0-1.1.49-1.1 1.1s.49 1.1 1.1 1.1c.61 0 1.1-.49 1.1-1.1s-.49-1.1-1.1-1.1zM12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm2.19 12.19L6 18l3.81-8.19L18 6l-3.81 8.19z"
            />
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
          </svg>
          <span
            className="MuiBottomNavigationAction-label-217"
          >
            Explore
          </span>
        </span>
        <span
          className="MuiTouchRipple-root-218"
        />
      </button>
    </div>
  </div>
</div>
`;
